src/biokbase/narrative/upa.py

```diff
@@ -0,0 +1,74 @@
+"""
+This is a reasonably tiny API for serializing and deserializing UPAs for storage in Narrative
+documents.
+
+"""
+
+import re
+from app_util import (system_variable)
+
+external_tag = "&"
+
+
+def is_upa(upa):
+    """
+    Returns True if the given upa string is valid, False, otherwise.
+    """
+    return re.match("^\d+(\/\d+){2}(;\d+(\/\d+){2})*$", upa) is not None
+
+
+def serialize(upa):
+    """
+    Serializes an UPA - prepares it for storage as a part of Narrative cell metadata.
+    This means a bit of a tweak to the UPA itself. Currently, we want to store it in a way
+    that designates it as a serialized string, and gives an easy path to substitute the
+    initial workspace part of the UPA with a different workspace.
+    So it gets transformed from:
+    ws1/obj1/ver1;ws2/obj2/ver2;...
+    to
+    [ws1]/obj1/ver1;ws2/obj2/ver2;...
+
+    In the case of UPAs representing objects that are located in a different workspace all
+    together (e.g. set items that aren't copied into the Narrative with the set container
+    object), they get flagged with a special character. In that case, the UPA is maintained,
+    but transformed into:
+    &ws1/obj1/ver1;ws2/obj2/ver2;...
+    """
+    if type(upa) is list:
+        upa = ";".join(upa)
+    if not is_upa(upa):
+        raise ValueError('"{}" is not a valid UPA. It may have already been serialized.'
+                         .format(upa))
+    ws_id = system_variable("workspace_id")
+    if ws_id is None:
+        raise RuntimeError("Currently loaded workspace is unknown! Unable to serialize UPA.")
+    ws_id = str(ws_id)
+    if upa.startswith(ws_id):
+        return upa.replace(ws_id, "[{}]".format(ws_id), 1)
+    else:
+        return external_tag + upa
```

Comment:
One thing that's a little worrisome is that this makes it easy for bugs to creep in by devs not being careful about their inputs. I'd almost want to split this into two methods - serialize_upa and serialize_external_upa to ensure the devs know exactly what they're doing when then serialize something.

Reply:
We talked about this a little bit, so just to record it here - the intent of this API is that it should be used by wrapping widgets and under-the-covers processes. Devs who write viewer widgets or apps will probably never interact with it.

But, you're right, we should be explicit. I'll make the change to split it.